k2/csrc/reverse.cu

```diff
@@ -0,0 +1,363 @@
+          // The 'idx0' must corresponds to a non-empty FSA as it is computed
+          // from the row_ids.
+          int32_t idx0 = fsas_row_ids1_data[idx01],
+                  idx0x = fsas_row_splits1_data[idx0],
+                  idx0x_final_state = fsas_row_splits1_data[idx0 + 1] - 1;
+          // Initialize as the offset.
+          int32_t new_idx01 = offset_idx01_data[idx0];
+          if (idx01 == idx0x) {
+            new_idx01 += idx0x_final_state;
+          } else if (idx01 == idx0x_final_state) {
+            new_idx01 += idx0x;
+          } else {
+            new_idx01 += idx01;
+          }
+          old2new_map_data[idx01] = new_idx01
+        });
+    return old2new_map;
+  }
+
+  
+  /*
+   Return the RaggedShape of the state-level reversed FsaVec.
+   Compared with the original FsaVec, we need to add an extra final state for
+   each non-empty FSA.
+   */
+  RaggedShape CreateReversedStateShape() {
+    Array1<int32_t> acc_non_empty_fsas = ExclusiveSum(non_empty_fsas_);
+    Array1<int32_t> reversed_row_split1 =
+      Plus(acc_non_empty_fsas, fsas_.RowSplits(1));
+    return RaggedShape2(&reversed_row_split1, nullptr, -1);
+  }
+
+
+  /*
+   Retrun the RaggedShape of the arc-level reversed FsaVec.
+   We generate the RaggedShape by figuring out the number of out-going arcs for
+   each reversed state. As the original source state will become to destination
+   state. So we collect the information by 'incoming_arcs'.
+
+   All cases will be one-to-one mapping, except that we need to add an extra
+   'final_arc' to the penultimate state of the reversed FSA (i.e. the
+   corresponding start state in original FSA).
+   */
+  RaggedShape CreateReversedArcSahpe(Ragged<int32_t> &incoming_arcs,
+                                     RaggedShape &reversed_states_shape) {
+    int32_t num_ori_states = fsas_.TotSize(1);
+    const int32_t *fsas_row_splits1_data = fsas_.RowSplits(1).Data(),
+                  *fsas_row_ids1_data = fsas_.RowIds(1).Data(), 
+                  *incoming_arcs_row_splits2_data =
+                    incoming_arcs.RowSplits(2).Data();
+
+    Array1<int32_t> num_arcs_per_reversed_state(
+        c_, reversed_states_shape.NumElements() + 1, 0);
+    int32_t *num_arcs_per_reversed_state_data =
+        num_arcs_per_reversed_state.Data();
+
+    // For each reversed FSA, we just need to add one final state which
+    // doesn't have the out-going arcs. So we deal with the each original state.
+    K2_EVAL(
+        c_, num_ori_states, lambda_set_num_arcs_per_reversed_state, 
+        (int32_t idx01)->void {
+          int32_t idx0 = fsas_row_ids1_data[idx01],
+                  idx0x = fsas_row_splits1_data[idx0],
+                  num_arcs = (idx01 == idx0x ? 1 : 0)
+          num_arcs += incoming_arcs_row_splits2_data[idx01 + 1] -
+                      incoming_arcs_row_splits2_data[idx01];
+          num_arcs_per_reversed_state[old2new_map_[idx01]] = num_arcs;
```

Comment:
```suggestion
          num_arcs_per_reversed_state_data[old2new_map_[idx01]] = num_arcs;
```

Also, please use `old2new_map_data[idx01]` in the kernel.
(You have to create `old2new_map_data` by yourself.